src/shop.c

```diff
@@ -0,0 +1,1283 @@
+                "\tadds r4, r1, 0\n"
+                "\tb _0809C0D4\n"
+                "\t.align 2, 0\n"
+                "_0809C0B8: .4byte gShopMenuHistory\n"
+                "_0809C0BC:\n"
+                "\tldrb r0, [r1, 0x16]\n"
+                "\tcmp r0, r5\n"
+                "\tbne _0809C0C8\n"
+                "\tadds r4, r1, 0\n"
+                "\tadds r4, 0xC\n"
+                "\tb _0809C0D4\n"
+                "_0809C0C8:\n"
+                "\tadds r4, r1, 0\n"
+                "\tadds r4, 0xC\n"
+                "\tcmp r2, 0\n"
+                "\tbne _0809C0D2\n"
+                "\tadds r4, r1, 0\n"
+                "_0809C0D2:\n"
+                "\tstrb r5, [r4, 0xA]\n"
+                "_0809C0D4:\n"
+                "\tldrh r0, [r4, 0x4]\n"
+                "\tcmp r0, 0\n"
+                "\tbeq _0809C0DE\n"
+                "\tmovs r0, 0x1\n"
+                "\tstrb r0, [r4, 0x9]\n"
+                "_0809C0DE:\n"
+                "\tstrh r3, [r4, 0x4]\n"
+                "\tldrh r1, [r4, 0x6]\n"
+                "\tldr r0, _0809C128 @ =0x000003e6\n"
+                "\tcmp r1, r0\n"
+                "\tbhi _0809C0F8\n"
+                "\tadds r0, r6, r1\n"
+                "\tstrh r0, [r4, 0x6]\n"
+                "\tlsls r0, 16\n"
+                "\tlsrs r0, 16\n"
+                "\tldr r1, _0809C12C @ =0x000003e7\n"
+                "\tcmp r0, r1\n"
+                "\tbls _0809C0F8\n"
+                "\tstrh r1, [r4, 0x6]\n"
+                "_0809C0F8:\n"
+                "\tldr r1, [r4]\n"
+                "\tldr r0, _0809C130 @ =0x000f423e\n"
+                "\tcmp r1, r0\n"
+                "\tbhi _0809C120\n"
+                "\tadds r0, r3, 0\n"
+                "\tbl itemid_get_market_price\n"
+                "\tlsls r0, 16\n"
+                "\tlsrs r0, 16\n"
+                "\tsubs r1, r5, 0x1\n"
+                "\tasrs r0, r1\n"
+                "\tadds r1, r0, 0\n"
+                "\tmuls r1, r6\n"
+                "\tldr r0, [r4]\n"
+                "\tadds r0, r1\n"
+                "\tstr r0, [r4]\n"
+                "\tldr r1, _0809C134 @ =0x000f423f\n"
+                "\tcmp r0, r1\n"
+                "\tbls _0809C120\n"
+                "\tstr r1, [r4]\n"
+                "_0809C120:\n"
+                "\tpop {r4-r6}\n"
+                "\tpop {r0}\n"
+                "\tbx r0\n"
+                "\t.align 2, 0\n"
+                "_0809C128: .4byte 0x000003e6\n"
+                "_0809C12C: .4byte 0x000003e7\n"
+                "_0809C130: .4byte 0x000f423e\n"
+                "_0809C134: .4byte 0x000f423f\n");
+}
+#endif
+
+static void RecordQuestLogItemPurchase(void)
+{
+    struct MartHistory *history = &gShopMenuHistory;
+    u16 v;
+
+    v = history->unkA;
+    if (v != 0)
+        sub_8113550(v + 0x24, (const u16*)history);
+    
+    v = history->unk16;
+    if (v != 0)
+    {
+        v += 0x24;
+        sub_8113550(v, (const u16*)&history->unkC);
```

Comment:
1) fix this after you change the type of `gShopMenuHistory`
2) not sure if `history` is really needed